File: src/main/java/com/freya02/botcommands/internal/application/slash/autocomplete/AutocompleteCommandParameter.java
package com.freya02.botcommands.internal.application.slash.autocomplete;

import com.freya02.botcommands.api.BContext;
import com.freya02.botcommands.api.application.CommandPath;
import com.freya02.botcommands.api.application.slash.autocomplete.annotations.CompositeKey;
import com.freya02.botcommands.api.parameters.SlashParameterResolver;
import com.freya02.botcommands.internal.application.slash.ApplicationCommandVarArgParameter;
import net.dv8tion.jda.api.entities.Member;
import net.dv8tion.jda.api.entities.Role;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.entities.channel.Channel;

import java.lang.reflect.Parameter;

public class AutocompleteCommandParameter extends ApplicationCommandVarArgParameter<SlashParameterResolver> {
	private final boolean compositeKey;

	public AutocompleteCommandParameter(BContext context, CommandPath path, Parameter parameter, int index) {
		super(context, path, SlashParameterResolver.class, parameter, index);

		if (User.class.isAssignableFrom(getBoxedType())
				|| Member.class.isAssignableFrom(getBoxedType())
				|| Channel.class.isAssignableFrom(getBoxedType())
				|| Role.class.isAssignableFrom(getBoxedType())) {
			throw new IllegalArgumentException("Autocomplete parameters cannot have an entity (User/Member/Channel/Role) as a value");
		}

		this.compositeKey = parameter.isAnnotationPresent(CompositeKey.class);
	}

	public boolean isCompositeKey() {
		return compositeKey;
	}
}

File: src/main/kotlin/com/freya02/botcommands/internal/application/slash/autocomplete/AutocompleteCommandParameter.kt
package com.freya02.botcommands.internal.application.slash.autocomplete

import com.freya02.botcommands.annotations.api.application.slash.autocomplete.annotations.CompositeKey
import com.freya02.botcommands.api.application.builder.SlashCommandOptionBuilder
import com.freya02.botcommands.api.parameters.SlashParameterResolver
import com.freya02.botcommands.internal.application.slash.AbstractSlashCommandParameter
import net.dv8tion.jda.api.entities.IMentionable
import kotlin.reflect.KParameter
import kotlin.reflect.full.hasAnnotation
import kotlin.reflect.full.isSubclassOf
import kotlin.reflect.jvm.jvmErasure

class AutocompleteCommandParameter(
    parameter: KParameter,
    optionBuilder: SlashCommandOptionBuilder, //TODO Not sure about that
    resolver: SlashParameterResolver<*, *>
) : AbstractSlashCommandParameter(
    parameter, optionBuilder, resolver
) {
    val isCompositeKey = parameter.hasAnnotation<CompositeKey>()

    init {
        if (type.jvmErasure.isSubclassOf(IMentionable::class)) {
            throw IllegalArgumentException("Autocomplete parameters cannot have an entity (anything extending IMentionable) as a value")
        }
    }
}